proposals/Exceptions.md

```diff
@@ -262,130 +250,471 @@ Like the `block`, `loop`, and `if` instructions, the `try` instruction is a
+which describe their use of the operand stack.
+
+Note that the textual `catch_all` instruction is implemented using the `else`
+operator. Since the `else` operator is always unambiguous in the binary format,
+there is no need to tie up a separate opcode for the `catch_all` instruction.
+
+
+---
+
+## Second Proposal - Instruction Part
+
+Exception handling allows code to break control flow when an exception is
+thrown. The exception can be any exception known by the WebAssembly module, or
+it may an unknown exception that was thrown by a called imported function.
+
+One of the problems with exception handling is that both WebAssembly and an
+embedder have different notions of what exceptions are, but both must be aware
+of the other.
+
+It is difficult to define exceptions in WebAssembly because (in general) it
+doesn't have knowledge of any embedder. Further, adding such knowledge to
+WebAssembly would limit the ability for other embedders to support WebAssembly
+exceptions.
+
+One issue is that both sides need to know if an exception was thrown by the
+other, because cleanup may need to be performed.
+
+Another problem is that WebAssembly doesn't have direct access to an embedder's
+memory. As a result, WebAssembly defers the handling of exceptions to the host
+VM.
+
+To access exceptions, WebAssembly provides instructions to check if the
+exception is one that WebAssembly understands. If so, the data of the
+WebAssembly exception is extracted and copied onto the stack, allowing
+succeeding instructions to process the data.
+
+Lastly, exception lifetimes may be maintained by the embedder, so that it can
+collect and reuse the memory used by exceptions. This implies that an embedder
+needs to know where exceptions are stored, so that it can determine when an
+exception can be garbage collected.
+
+This also implies that that embedders must provide a garbage collector for
+exceptions. For embedders that have garbage collection (such as JavaScript),
+this is not a problem.
+
+However, not all embedders may have a garbage collector. For this reason,
+WebAssembly exceptions are designed to allow other storage management methods,
+such as reference counting, to perform the garbage collection in the embedder.
+
+To do this, WebAssembly exceptions are immutable once created, to avoid cyclic
+data structures that cannot be garbage collected. It also means that exceptions
+can't be stored into linear memory. The rationale for this is twofold:
+
+* For security. Loads and stores do not guarantee that the data read was of the
+  same type as stored. This allows spoofing of exception references that may
+  allow a WebAssembly module to access data it should not know in the host VM.
+
+* The embedder does not know the layout of data in linear memory, so it can't
+  find places where exception references are stored.
+
+Hence, while an exception reference is a new first class type, this proposal
```

Comment:
Perhaps include a link to the reference-types proposal here and clarify the intention that exceptions are added as an additional reference type.